From d9271d747ef42944b37dfce303795d1df8c22587 Mon Sep 17 00:00:00 2001
From: Alistair Leslie-Hughes <leslie_alistair@hotmail.com>
Date: Mon, 22 Nov 2021 16:17:01 +1100
Subject: [PATCH] msdasql: HACK - IColumnsInfo GetColumnInfo for ICommandText

This needs to return columns.

This patch shows that ICommandPrepare_Prepare needs to be implement
by using SQLPrepare with a valid connection.  Then we can query the
columns without issues.  We should cache these values, as they aren't
going to change between GetColumnInfo(Command), and GetColumnInfo(Rowset).
---
 dlls/msdasql/session.c        | 15 +++++-
 dlls/msdasql/tests/provider.c | 90 +++++++++++++++++++++++------------
 2 files changed, 73 insertions(+), 32 deletions(-)

diff --git a/dlls/msdasql/session.c b/dlls/msdasql/session.c
index 3247b6aa664..5a414ca5d2f 100644
--- a/dlls/msdasql/session.c
+++ b/dlls/msdasql/session.c
@@ -1782,7 +1782,20 @@ static HRESULT WINAPI colsinfo_GetColumnInfo(IColumnsInfo *iface, DBORDINAL *col
 {
     struct command *command = impl_from_IColumnsInfo( iface );
     FIXME("%p, %p, %p, %p\n", command, columns, colinfo, stringsbuffer);
-    return E_NOTIMPL;
+
+    if (!command->query)
+        return DB_E_NOCOMMAND;
+
+    if (columns)
+        *columns = 0;
+
+    if (colinfo)
+        *colinfo = NULL;
+
+    if (stringsbuffer)
+        *stringsbuffer = NULL;
+
+    return DB_E_NOTPREPARED;
 }
 
 static HRESULT WINAPI colsinfo_MapColumnIDs(IColumnsInfo *iface, DBORDINAL column_ids,
diff --git a/dlls/msdasql/tests/provider.c b/dlls/msdasql/tests/provider.c
index f7b9f426cb2..d29119a8a7d 100644
--- a/dlls/msdasql/tests/provider.c
+++ b/dlls/msdasql/tests/provider.c
@@ -69,6 +69,37 @@ static void free_dbpropinfoset(ULONG count, DBPROPINFOSET *propinfoset)
     CoTaskMemFree(propinfoset);
 }
 
+static char *dump_flags(DWORD flags)
+{
+    static char buffer[2048];
+
+    memset(buffer, 0, sizeof(buffer));
+    if(flags & DBCOLUMNFLAGS_ISBOOKMARK)
+        strcat(buffer, "DBCOLUMNFLAGS_ISBOOKMARK | ");
+    if(flags & DBCOLUMNFLAGS_MAYDEFER)
+        strcat(buffer, "DBCOLUMNFLAGS_MAYDEFER | ");
+    if(flags & DBCOLUMNFLAGS_WRITE)
+        strcat(buffer, "DBCOLUMNFLAGS_WRITE | ");
+    if(flags & DBCOLUMNFLAGS_WRITEUNKNOWN)
+        strcat(buffer, "DBCOLUMNFLAGS_WRITEUNKNOWN | ");
+    if(flags & DBCOLUMNFLAGS_ISFIXEDLENGTH)
+        strcat(buffer, "DBCOLUMNFLAGS_ISFIXEDLENGTH | ");
+    if(flags & DBCOLUMNFLAGS_ISNULLABLE)
+        strcat(buffer, "DBCOLUMNFLAGS_ISNULLABLE | ");
+    if(flags & DBCOLUMNFLAGS_MAYBENULL)
+        strcat(buffer, "DBCOLUMNFLAGS_MAYBENULL | ");
+    if(flags & DBCOLUMNFLAGS_ISLONG)
+        strcat(buffer, "DBCOLUMNFLAGS_ISLONG | ");
+    if(flags & DBCOLUMNFLAGS_ISROWID)
+        strcat(buffer, "DBCOLUMNFLAGS_ISROWID | ");
+    if(flags & DBCOLUMNFLAGS_ISROWVER)
+        strcat(buffer, "DBCOLUMNFLAGS_ISROWVER | ");
+    if(flags & DBCOLUMNFLAGS_CACHEDEFERRED)
+        strcat(buffer, "DBCOLUMNFLAGS_CACHEDEFERRED");
+
+    return buffer;
+}
+
 static void test_msdasql(void)
 {
     HRESULT hr;
@@ -444,6 +475,11 @@ static void test_command_text(IUnknown *cmd)
     HRESULT hr;
     OLECHAR *str;
     GUID dialect;
+    IColumnsInfo *colinfo;
+    ICommandPrepare *commandprepare;
+    DBORDINAL columns;
+    DBCOLUMNINFO *dbcolumninfo;
+    OLECHAR *buffer;
 
     hr = IUnknown_QueryInterface(cmd, &IID_ICommandText, (void**)&command_text);
     ok(hr == S_OK, "got 0x%08lx\n", hr);
@@ -451,6 +487,12 @@ static void test_command_text(IUnknown *cmd)
     hr = ICommandText_GetCommandText(command_text, &dialect, &str);
     ok(hr == DB_E_NOCOMMAND, "got 0x%08lx\n", hr);
 
+    hr = IUnknown_QueryInterface(cmd, &IID_IColumnsInfo, (void**)&colinfo);
+    ok(hr == S_OK, "got 0x%08lx\n", hr);
+
+    hr = IColumnsInfo_GetColumnInfo(colinfo, &columns, &dbcolumninfo, &buffer);
+    ok(hr == DB_E_NOCOMMAND, "got 0x%08lx\n", hr);
+
 if (0)
 {
     /* Crashes under windows */
@@ -487,6 +529,23 @@ if (0)
     ok (!lstrcmpW(L"select * from testing", str), "got %s\n", debugstr_w(str));
     HeapFree(GetProcessHeap(), 0, str);
 
+    hr = ICommandText_QueryInterface(command_text, &IID_ICommandPrepare, (void**)&commandprepare);
+    ok(hr == S_OK, "got 0x%08lx\n", hr);
+
+    ok (0, "*************\n");
+    hr = ICommandPrepare_Prepare(commandprepare, 1);
+    ok(hr == S_OK, "got 0x%08lx\n", hr);
+    ok (0, " *************\n");
+
+    hr = IColumnsInfo_GetColumnInfo(colinfo, &columns, &dbcolumninfo, &buffer);
+    ok(hr == S_OK, "got 0x%08lx\n", hr);
+    ok(columns == 3, "got %ld\n", columns);
+
+    CoTaskMemFree(dbcolumninfo);
+    CoTaskMemFree(buffer);
+
+    ICommandPrepare_Release(commandprepare);
+    IColumnsInfo_Release(colinfo);
     ICommandText_Release(command_text);
 }
 
@@ -506,37 +565,6 @@ static void test_command_dbsession(IUnknown *cmd, IUnknown *session)
     ICommandText_Release(command_text);
 }
 
-static char *dump_flags(DWORD flags)
-{
-    static char buffer[2048];
-
-    memset(buffer, 0, sizeof(buffer));
-    if(flags & DBCOLUMNFLAGS_ISBOOKMARK)
-        strcat(buffer, "DBCOLUMNFLAGS_ISBOOKMARK | ");
-    if(flags & DBCOLUMNFLAGS_MAYDEFER)
-        strcat(buffer, "DBCOLUMNFLAGS_MAYDEFER | ");
-    if(flags & DBCOLUMNFLAGS_WRITE)
-        strcat(buffer, "DBCOLUMNFLAGS_WRITE | ");
-    if(flags & DBCOLUMNFLAGS_WRITEUNKNOWN)
-        strcat(buffer, "DBCOLUMNFLAGS_WRITEUNKNOWN | ");
-    if(flags & DBCOLUMNFLAGS_ISFIXEDLENGTH)
-        strcat(buffer, "DBCOLUMNFLAGS_ISFIXEDLENGTH | ");
-    if(flags & DBCOLUMNFLAGS_ISNULLABLE)
-        strcat(buffer, "DBCOLUMNFLAGS_ISNULLABLE | ");
-    if(flags & DBCOLUMNFLAGS_MAYBENULL)
-        strcat(buffer, "DBCOLUMNFLAGS_MAYBENULL | ");
-    if(flags & DBCOLUMNFLAGS_ISLONG)
-        strcat(buffer, "DBCOLUMNFLAGS_ISLONG | ");
-    if(flags & DBCOLUMNFLAGS_ISROWID)
-        strcat(buffer, "DBCOLUMNFLAGS_ISROWID | ");
-    if(flags & DBCOLUMNFLAGS_ISROWVER)
-        strcat(buffer, "DBCOLUMNFLAGS_ISROWVER | ");
-    if(flags & DBCOLUMNFLAGS_CACHEDEFERRED)
-        strcat(buffer, "DBCOLUMNFLAGS_CACHEDEFERRED");
-
-    return buffer;
-}
-
 static void test_rowset_interfaces(IRowset *rowset, ICommandText *commandtext)
 {
     IRowsetInfo *info;
-- 
2.42.0

